### PR TITLE
feat: Support XUDT

### PR DESCRIPTION
**Pull request by yanguoyu**
Refer to https://github.com/Magickbase/neuron-public-issues/issues/379

https://github.com/Magickbase/neuron-public-issues/issues/379#issuecomment-2209904504 is the test list

Asset Account:

https://github.com/nervosnetwork/neuron/assets/12881040/7ee37205-b25e-4566-8498-513b38a149a1

Customized Assets:

https://github.com/nervosnetwork/neuron/assets/12881040/7207d91d-4e94-49f1-ad13-c5551da8d9f7

History, Overview And CellManage:

https://github.com/nervosnetwork/neuron/assets/12881040/1627124d-80c5-47fc-bd92-1bc7d2dcb149

### CODE REVIEW

**Comment by Danie0918** [comment]
@devchenyan @homura Please have a review.

File: packages/neuron-ui/src/components/CellManagement/hooks.ts [hunk 1]
```diff
@@ -18,7 +18,8 @@ import { ErrorCode, LockScriptCategory, RoutePath, TypeScriptCategory, isSuccess
 import { SortType } from 'widgets/Table'
 
 const cellTypeOrder: Record<string, number> = {
-  [TypeScriptCategory.SUDT]: 1,
+  [TypeScriptCategory.SUDT]: 0,
+  [TypeScriptCategory.XUDT]: 1,
```
[the remark this answers is not in the document]
Comment (yanguoyu) [reply packages/neuron-ui/src/components/CellManagement/hooks.ts:22]: This value is used to order the `type script`.

File: packages/neuron-wallet/src/database/chain/migrations/1720089814860-AddUdtType.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import {MigrationInterface, QueryRunner} from "typeorm";
+
+export class AddUdtType1720089814860 implements MigrationInterface {
+    name = 'AddUdtType1720089814860'
+
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        await queryRunner.query(`ALTER TABLE "sudt_token_info" ADD COLUMN "udtType" varchar;`)
+        await queryRunner.query(`CREATE UNIQUE INDEX "IDX_37bb4f2a4a849bf0b1dadee2c7" ON "sudt_token_info" ("tokenID", "udtType") `);
```
[the remark this answers is not in the document]
Comment (yanguoyu) [reply packages/neuron-wallet/src/database/chain/migrations/1720089814860-AddUdtType.ts:8]: It was created by the `typeorm` command.

[the remark this answers is not in the document]
Comment (homura) [reply packages/neuron-wallet/src/database/chain/migrations/1720089814860-AddUdtType.ts:8]: got it

**Review by devchenyan — APPROVED** [review]

**Comment by Keith-CY** [comment]
@homura please check

**Review by homura — APPROVED** [review]

**Comment by silySuper** [comment]
I have known that xudt does not have tool like this https://ckb.tools/sudt,so it will try to use code to edit and transfer xudt.

**Comment by silySuper** [comment]
1.In history page,when I received 1,000  XTT21 https://pudge.explorer.nervos.org/transaction/0x57132e618c65a2c665ee72b199c2804c0379eefb8885a3c6e8028a614e1549f9?layout=lite
neuron asset shows --(both in overview page):
截屏2024-07-17 15 07 19
截屏2024-07-17 15 50 46

Does here need show  ckb  asset?  @Danie0918 

In customized asset,asset shows 100000000000(has no unit) ,Does asset not show 1000?@Danie0918 
截屏2024-07-17 15 38 20
In asset account ,it shows:

截屏2024-07-17 15 40 53

when I want to change account info,it can not confirm.

截屏2024-07-17 15 43 02

**Comment by yanguoyu** [comment]
@silySuper Before you update the asset info, Neuron doesn't know the token information. So it can only show the full mount without decimal. After the token information is updated, it will show normal.

Under normal circumstances, token information can be saved correctly. Could you describe the reproducible steps for the inability to edit?

image

**Comment by silySuper** [comment]
2.Send xudt from another wallet(wallet 9 --->  test01 xudt account address),UI choose is only one ,and when click send shows an error.
截屏2024-07-19 10 49 16
截屏2024-07-19 10 50 00

**Comment by yanguoyu** [comment]
> 2.Send xudt from another wallet(wallet 9 ---> test01 xudt account address),UI choose is only one ,and when click send shows an error. 截屏2024-07-19 10 49 16 截屏2024-07-19 10 50 00

Because the address that you sent to is not a secp256k1 address, it does not support sending with the `cheque` lock. Could you paste the send address and upload the log?

**Comment by silySuper** [comment]
> > 2.Send xudt from another wallet(wallet 9 ---> test01 xudt account address),UI choose is only one ,and when click send shows an error. 截屏2024-07-19 10 49 16 截屏2024-07-19 10 50 00
> 
> Because the address that you sent to is not a secp256k1 address, it does not support sending with the `cheque` lock. Could you paste the send address and upload the log?

address is:
ckt1qq6pngwqn6e9vlm92th84rk0l4jp2h8lurchjmnwv8kq3rt5psf4vq0hmqz5aayuhd5ar23wyn3ncstacrs07wc3e4hzz
截屏2024-07-19 11 29 51

**Comment by yanguoyu** [comment]
> ckt1qq6pngwqn6e9vlm92th84rk0l4jp2h8lurchjmnwv8kq3rt5psf4vq0hmqz5aayuhd5ar23wyn3ncstacrs07wc3e4hzz

/package fixed

**Comment by silySuper** [comment]
3.Sudt send to xudt address (ckt1qq6pngwqn6e9vlm92th84rk0l4jp2h8lurchjmnwv8kq3rt5psf4vqvz8rmuhallzq490rp2xk3fl29vhyzmngq9ej7sm), .After send success sudt asset is decreased,but xudt does not increase.


https://github.com/user-attachments/assets/ba08b64a-14fa-4a71-9922-43467df92570

**Comment by yanguoyu** [comment]
> 3.Sudt send to xudt address (ckt1qq6pngwqn6e9vlm92th84rk0l4jp2h8lurchjmnwv8kq3rt5psf4vqvz8rmuhallzq490rp2xk3fl29vhyzmngq9ej7sm), .After send success sudt asset is decreased,but xudt does not increase.
> 
>  2024-07-19.15.38.20.mov

I guess you have two sudt accounts for this address.
https://pudge.explorer.nervos.org/address/ckt1qq6pngwqn6e9vlm92th84rk0l4jp2h8lurchjmnwv8kq3rt5psf4vqvz8rmuhallzq490rp2xk3fl29vhyzmngq9ej7sm
From the explorer: `0x641238cd95b9f5e6224da1963a0bcaa2e972dc87e56eb1ded9a35dcebbc37ff4` only have 10011 amount.
And it has another sudt `0xb1718e7c0175d8a6428a6ddca708b765803e3131d07c8e0046a94be310f1722b` with amount  320000 
image

**Comment by silySuper** [comment]
1.verified 2.verified 3.not  need to change

**Review by silySuper — APPROVED** [review]

**Pull request merged**